fiaas_deploy_daemon/crd/apiextensionsv1_crd_watcher.py

```diff
@@ -0,0 +1,220 @@
+# coding: utf-8
+
+# Copyright 2017-2019 The FIAAS Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+import logging
+
+from k8s.base import WatchEvent
+from k8s.client import NotFound
+from k8s.models.common import ObjectMeta
+from k8s.models.apiextensions_v1_custom_resource_definition import CustomResourceConversion,\
+    CustomResourceDefinitionNames, CustomResourceDefinitionSpec, CustomResourceDefinition,\
+    CustomResourceDefinitionVersion, CustomResourceValidation, JSONSchemaProps
+from k8s.watcher import Watcher
+from yaml import YAMLError
+
+from .status import create_name
+from .types import FiaasApplication, FiaasApplicationStatus
+from ..base_thread import DaemonThread
+from ..deployer import DeployerEvent
+from ..log_extras import set_extras
+from ..retry import retry_on_upsert_conflict
+from ..specs.factory import InvalidConfiguration
+
+LOG = logging.getLogger(__name__)
+
+
+class ApiextensionsV1CrdWatcher(DaemonThread):
+    def __init__(self, spec_factory, deploy_queue, config, lifecycle):
+        super(ApiextensionsV1CrdWatcher, self).__init__()
+        self._spec_factory = spec_factory
+        self._deploy_queue = deploy_queue
+        self._watcher = Watcher(FiaasApplication)
+        self._lifecycle = lifecycle
+        self.namespace = config.namespace
+        self.enable_deprecated_multi_namespace_support = config.enable_deprecated_multi_namespace_support
+
+    def __call__(self):
+        while True:
+            if self.enable_deprecated_multi_namespace_support:
+                self._watch(namespace=None)
+            else:
+                self._watch(namespace=self.namespace)
+
+    def _watch(self, namespace):
+        try:
+            for event in self._watcher.watch(namespace=namespace):
+                self._handle_watch_event(event)
+        except NotFound:
+            self.create_custom_resource_definitions()
```

Comment:
When upgrading fiaas-deploy-daemon to this version and enabling the `--use-apiextensionsv1-crd` flag in an existing cluster, the CRDs won't be updated, since they already exist as apiextensions/v1beta1 CRDs and are served via the apiextensions/v1 api also. It seems that skipper actually always updates the CRDs when it starts (see [1] and [2]), so fiaas-deploy-daemon perhaps should do the same? 

Out of necessity CRDs are managed in both skipper an fiaas-deploy-daemon. This means that operators must make sure to upgrade and enable CRD v1 in both skipper and fiaas-deploy-daemon at the same time, otherwise these will try to enforce different versions of the CRDs. When skipper is no longer in use (#163) fiaas-deploy-daemon (only) will own the CRDs, which will remove the potential for this situation.

[1]: https://github.com/fiaas/skipper/blob/8741ab0d8b25ac5c65bf21407f991dcdb3f31cac/fiaas_skipper/__init__.py#L120
[2]: https://github.com/fiaas/skipper/blob/8741ab0d8b25ac5c65bf21407f991dcdb3f31cac/fiaas_skipper/deploy/crd/bootstrap.py#L46

Reply:
Changed it to create/update CRD resources every time a CRD Watcher is started